Reorganise qa directory

jnewbery opened the pull request:
Initially proposed by @jonasschnelli here: https://botbot.me/freenode/bitcoin-core-dev/2017-02-16/?msg=81095853&page=5

This re-organises the `qa` test directory to meet the following requirements:

- qa directory should be renamed to come after /src alpabetically, so that git diffs and github PRs show changes to qa tests beneath changes to product code.
- rpc-tests should be renamed to something, since they test more than the RPC interface
- pull-tester should be renamed to something, since it's run locally as well as on PRs
- bitcoin-util-test should be moved to the qa test directory, since it's a functional test rather than a unit test.

This PR has the following commits:

- commit 1 renames qa to test
- commit 2 renames the rpc-tests directory to qa
- commit 3 renames test/pull-tester/rpc-tests.py to test/runner/runner.py
- commit 4 renames the --enable-extended-rpc-tests configure option to --enable-extended-qa-tests
- commit 5 moves the bitcoin-util-test.py test into test/util

The new directory structure is [**EDITED**]:

```
|--src
|   |--<source files etc>
|   |--test
|       |---<unit tests>
|
|--test
   |---functional
   |   |---<tests (were previously in /qa/rpc-tests)>
   |   |---test_runner.py (was previously qa/pull-tester/rpc-tests.py)
   |   |---test_config.ini
   |  
   |---README.md
   |
   |---util
   |   |---bctest.py (was previously in src/test)
   |   |---bitcoin-util-test.py (was previously in src/test)
   |   |---buildenv.py.in (was previously in src/test)
   |   |---data (was previously in src/test)
   |   |   |---<data files>
```

The actual structure within the new directories is unchanged to minimize changes.

I've tested this at each commit by rebuilding and running the test cases. Reviewers note that you'll need to ./autogen.sh and ./configure to test this.

Any feedback on names or structure welcomed.

jnewbery committed:
Rename qa directory to test

jnewbery committed:
Move src/test/bitcoin-util-test.py to test/util/bitcoin-util-test.py

jonasschnelli commented:
Concept ACK.
What about using ...
```
|--test
   |---rpc-tests
```
instead of
```
|--test
   |---qa
```
?

sipa commented:
@jonasschnelli It doesn't contain purely RPC tests anymore (also P2P and REST).

@jnewberry Why is runner separate? My #1 complaint about the current structure is that the runner is in a separate directory.

laanwj commented:
Concept ACK. New structure looks good to me.

Agree that there's no need for the runner to be separate.

jonasschnelli commented:
> Indeed. These are "functional tests" or "integration tests" in contrast to the "unit tests" which are part of the C++ source code. They make use of RPC, but that doesn't define them, ideally they should test every interface.

Okay. That makes sense.

@jnewbery: I think you need to update `.travis.yml`.
Tested a bit and it seems to work well.

jnewbery committed:
Rename rpc-tests directory to functional

jnewbery committed:
Rename test/pull-tester/rpc-tests.py to test/functional/test_runner.py

jnewbery committed:
Rename --enable-extended-rpc-tests to --enable-extended-functional-tests

jnewbery commented:
Thanks everyone for the feedback. I've made the following changes:

- commit 1 renames qa to test
- commit 2 renames the rpc-tests directory to ~qa~ **functional**
- commit 3 renames test/pull-tester/rpc-tests.py to ~test/runner/runner.py~ **test/functional/test_runner.py and renames test/pull-tester/tests_config.ini.in to test/functional/config.ini.in. It also updates .travis.yml so travis knows where to find test_runner.py**
- commit 4 renames the --enable-extended-rpc-tests configure option to ~--enable-extended-qa-tests~ **--enable-extended-functional-tests**
- commit 5 moves the bitcoin-util-test.py test into test/util

also rebased.

jnewbery commented:
rebased

fanquake commented:
Needs .gitignore updated to include `test/functional/config.ini` and `test/util/buildenv.py`. You can probably drop the old `qa/pull-tester/tests_config.py` entry now as well.

Currently failing with:
```
==============================================
   Bitcoin Core 0.14.99: src/test-suite.log
==============================================

# TOTAL: 2
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test/test_bitcoin
=======================

Assertion failed: lock cs_wallet not held in wallet/wallet.cpp:163; locks held:
Running 232 test cases...
unknown location:0: fatal error: in "wallet_tests/rescan": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp:362: last checkpoint: "rescan" entry.
Assertion failed: (secp256k1_context_sign == NULL), function ECC_Start, file key.cpp, line 300.
unknown location:0: fatal error: in "wallet_tests/coin_mark_dirty_immature_credit": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp:435: last checkpoint: "coin_mark_dirty_immature_credit" fixture entry.
Assertion failed: (secp256k1_context_sign == NULL), function ECC_Start, file key.cpp, line 300.
unknown location:0: fatal error: in "wallet_tests/ComputeTimeSmart": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp:479: last checkpoint: "ComputeTimeSmart" fixture entry.
Assertion failed: (secp256k1_context_sign == NULL), function ECC_Start, file key.cpp, line 300.
unknown location:0: fatal error: in "wallet_crypto/passphrase": signal: SIGABRT (application abort requested)
wallet/test/crypto_tests.cpp:186: last checkpoint: "passphrase" fixture entry.
Assertion failed: (secp256k1_context_sign == NULL), function ECC_Start, file key.cpp, line 300.
unknown location:0: fatal error: in "wallet_crypto/encrypt": signal: SIGABRT (application abort requested)
wallet/test/crypto_tests.cpp:202: last checkpoint: "encrypt" fixture entry.
Assertion failed: (secp256k1_context_sign == NULL), function ECC_Start, file key.cpp, line 300.
unknown location:0: fatal error: in "wallet_crypto/decrypt": signal: SIGABRT (application abort requested)
wallet/test/crypto_tests.cpp:217: last checkpoint: "decrypt" fixture entry.

*** 6 failures are detected in the test module "Bitcoin Test Suite"
Assertion failed: (!pthread_mutex_destroy(&m)), function ~recursive_mutex, file /usr/local/include/boost/thread/pthread/recursive_mutex.hpp, line 104.
FAIL test/test_bitcoin (exit status: 134)
```

sipa commented:
Needs rebase.

theuni commented:
@jnewbery Other than the rebase, is this now ready for review? Travis is happy, at least.

jnewbery commented:
@theuni thanks for your help on this. I've rebased and squashed, so assuming travis passes, this is now ready for review. The two parts which I'm a little unsure of and I'd especially like reviewers to look at are:

- changes to /contrib/rpm/bitcoin.spec. I actually think we can remove the line:
```srcdir=. test/bitcoin-util-test.py```
entirely, since bitcoin-util-test.py is run as part of make check, but I've left it in for now. Let me know if you think we can remove it.
- changes to Makefile.am, configure.ac and src/test/Makefile.test.include in the final commit (Move src/test/bitcoin-util-test.py to test/util/bitcoin-util-test.py)

jonasschnelli commented:
Needs rebase.

MarcoFalke commented:
No need to rebase this until the merge is scheduled. Needs review first.

On Thu, Mar 16, 2017 at 2:33 PM, Jonas Schnelli <notifications@github.com>
wrote:

> Needs rebase.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9956#issuecomment-287057726>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmvyo8dHQFuV1SI6ibAgsXiBQCj7_Iks5rmToygaJpZM4MXbTb>
> .
>

theuni commented on test/functional/test_runner.py:
```diff
@@ -87,7 +86,7 @@
 ZMQ_SCRIPTS = [
     # ZMQ test can only be run if bitcoin was built with zmq-enabled.
-    # call rpc_tests.py with -nozmq to explicitly exclude these tests.
+    # call runner.py with -nozmq to explicitly exclude these tests.
```
runner -> test_runner from here down?

theuni commented on configure.ac:
```diff
@@ -1131,10 +1131,13 @@ AC_SUBST(EVENT_PTHREADS_LIBS)
-AC_CONFIG_FILES([Makefile src/Makefile doc/man/Makefile share/setup.nsi share/qt/Info.plist src/test/buildenv.py])
-AC_CONFIG_FILES([qa/pull-tester/tests_config.ini],[chmod +x qa/pull-tester/tests_config.ini])
+AC_CONFIG_FILES([Makefile src/Makefile doc/man/Makefile share/setup.nsi share/qt/Info.plist])
+AC_CONFIG_FILES([test/functional/config.ini],[chmod +x test/functional/config.ini])
```
Do these really need +x?

theuni commented on configure.ac:
```diff
@@ -414,8 +414,8 @@ if test x$use_pkgconfig = xyes; then
-if test x$use_extended_rpc_tests != xno; then
-  AC_SUBST(EXTENDED_RPC_TESTS, -extended)
+if test x$use_extended_functional_tests != xno; then
+  AC_SUBST(EXTENDED_FUNCTIONAL_TESTS, -extended)
```
Please update Makefile.am to match. I'm pretty sure the cov stuff is busted atm, but at least this part won't be busted too.

Actually, it'd make sense to use this in test/functional/config.ini.in directly if you'd prefer to skip to that.

MarcoFalke replied:
This should be `--extended` with two dashes?

theuni commented:
Looks good to me other than a few little nits. Passes Travis' out-of-tree-distdir gauntlet.

I'd like to move the related stuff to a new Makefile.include, but as a next step.

I'm unsure about the bitcoin.spec, but we probably can't assume they're running it from inside our tree?

jtimon commented:
Could you update the struct on the top instead of only the comments? It still says tests/runner and test/qa.
Concept ACK. Needs rebase.

I think "qa" instead of "functional" was just fine, but I guess "functional", or "integration" or "tests" or "py" would be fine as well. Maybe I just prefer qa because it's shorter. Anyway, bikeshedding.

About the runner and the config, what about just leaving them in the top level test wallet instead of putting them with the rest of the tests in functional?

jnewbery commented:
> Could you update the struct on the top instead of only the comments? It still says tests/runner and test/qa.

Apologies. Now updated.

> I think "qa" instead of "functional" was just fine, but I guess "functional", or "integration" or "tests" or "py" would be fine as well. Maybe I just prefer qa because it's shorter. Anyway, bikeshedding.

I think functional or integration are best because they are strongly differentiated from unit tests. qa is kind of meaningless in this context.

> About the runner and the config, what about just leaving them in the top level test wallet instead of putting them with the rest of the tests in functional?

I aim to eventually merge `test/functional/config.ini.in` and `test/util/buildenv.py` into a single config file which will live in the base `/test` directory. I think it makes sense to have `test_runner.py` live in the `/test/functional` directory since it only runs the functional tests. Having it in the base `/test` may confuse people into thinking it also runs the util tests.

jnewbery commented:
@theuni's comments addressed, and rebased.

Old tip: https://github.com/jnewbery/bitcoin/tree/pr9956.2 / https://github.com/jnewbery/bitcoin/commit/3d42c4e508ae0fa73eef5ae6584be3a624d5d129
new tip: https://github.com/jnewbery/bitcoin/tree/pr9956.3 / https://github.com/jnewbery/bitcoin/commit/f1653606a6c5803c4711ddaa4fea8242c78393e3 

Ready for additional review.

jtimon commented:
Random thought, maybe create a subdirectory for the extended tests? That would make it more clear I think.

fanquake commented:
I've been testing this, looks good so far 👍 . However, needs another rebase.

MarcoFalke commented on contrib/rpm/bitcoin.spec:
```diff
@@ -311,10 +311,8 @@ rm -f %{buildroot}%{_bindir}/test_*
-popd
-qa/pull-tester/rpc-tests.py -extended
+srcdir=src test/bitcoin-util-test.py
+test/functional/test_runner.py -extended
```
Also double dash?

MarcoFalke commented on test/functional/config.ini.in:
```diff
@@ -3,7 +3,7 @@
 
 # These environment variables are set by the build process and read by
-# rpc-tests.py
+# test/runner/runner.py
```
'test/runner/runner.py': No such file or directory

MarcoFalke commented on test/README.md:
```diff
@@ -1,5 +1,5 @@
-The [pull-tester](/qa/pull-tester/) folder contains a script to call
```
Can you mention `test/util/bitcoin-util-test.py` in this file as well?

maflcko reviewed (commented):
utACK f1653606a6c5803c4711ddaa4fea8242c78393e3, but needs feedback addressed

laanwj commented:
As soon as there is agreement on this we should probably merge this. Otherwise it will continue to be outdated by every single change in the tests.

MarcoFalke commented:
I plan to merge this after the next rebase, so if someone disagrees with anything, please raise your voice.

jnewbery commented:
I've addressed @MarcoFalke's comments and rebased. New tip: https://github.com/jnewbery/bitcoin/tree/pr9956.4 / https://github.com/jnewbery/bitcoin/commit/63d66ba20a634b54f6d5e8b051fb4a106f2cef6c

@jtimon - re: extended tests in a subdirectory: Interesting idea. I'm not convinced that it's necessary, but it's orthognal to this move, so let's not hold this PR up on that.

fanquake commented:
Merged and ran `test/functional/test_runner.py` with no arguments:
```
TEST                           | PASSED | DURATION

fundrawtransaction.py          | True   | 192 s
walletbackup.py                | True   | 311 s
p2p-compactblocks.py           | True   | 166 s
segwit.py                      | True   | 83 s
wallet-accounts.py             | True   | 65 s
wallet.py                      | True   | 149 s
wallet-dump.py                 | True   | 50 s
wallet-hd.py                   | True   | 574 s
p2p-segwit.py                  | True   | 116 s
zapwallettxes.py               | True   | 29 s
listtransactions.py            | True   | 53 s
p2p-fullblocktest.py           | True   | 613 s
sendheaders.py                 | True   | 49 s
mempool_limit.py               | True   | 16 s
merkle_blocks.py               | True   | 25 s
importmulti.py                 | True   | 40 s
abandonconflict.py             | True   | 23 s
receivedby.py                  | True   | 31 s
mempool_resurrect_test.py      | True   | 4 s
bip68-112-113-p2p.py           | True   | 29 s
reindex.py                     | True   | 19 s
txn_doublespend.py --mineblock | True   | 18 s
rawtransactions.py             | True   | 34 s
mempool_spendcoinbase.py       | True   | 4 s
txn_clone.py                   | True   | 19 s
mempool_reorg.py               | True   | 11 s
httpbasics.py                  | True   | 11 s
rest.py                        | True   | 21 s
multi_rpc.py                   | True   | 4 s
signrawtransactions.py         | True   | 5 s
decodescript.py                | True   | 4 s
getchaintips.py                | True   | 43 s
proxy_test.py                  | True   | 15 s
blockchain.py                  | True   | 7 s
disablewallet.py               | True   | 3 s
p2p-mempool.py                 | True   | 4 s
keypool.py                     | True   | 11 s
invalidblockrequest.py         | True   | 6 s
prioritise_transaction.py      | True   | 11 s
nodehandling.py                | True   | 26 s
invalidtxrequest.py            | True   | 6 s
signmessages.py                | True   | 5 s
importprunedfunds.py           | True   | 13 s
nulldummy.py                   | True   | 6 s
preciousblock.py               | True   | 16 s
p2p-versionbits-warning.py     | True   | 19 s
rpcnamedargs.py                | True   | 4 s
p2p-leaktests.py               | True   | 10 s
zmq_test.py                    | True   | 19 s
bumpfee.py                     | True   | 41 s
listsinceblock.py              | True   | 49 s
import-rescan.py               | True   | 59 s

ALL                            | True   | 3141 s (accumulated)

Runtime: 803 s
```

Running through the extended tests and passing some options now.

MarcoFalke commented:
utACK 63d66ba

and thanks @fanquake for testing!

On Mon, Mar 20, 2017 at 4:35 PM, Michael <notifications@github.com> wrote:

> Merged and ran test/functional/test_runner.py with no arguments:
>
> TEST                           | PASSED | DURATION
>
> fundrawtransaction.py          | True   | 192 s
> walletbackup.py                | True   | 311 s
> p2p-compactblocks.py           | True   | 166 s
> segwit.py                      | True   | 83 s
> wallet-accounts.py             | True   | 65 s
> wallet.py                      | True   | 149 s
> wallet-dump.py                 | True   | 50 s
> wallet-hd.py                   | True   | 574 s
> p2p-segwit.py                  | True   | 116 s
> zapwallettxes.py               | True   | 29 s
> listtransactions.py            | True   | 53 s
> p2p-fullblocktest.py           | True   | 613 s
> sendheaders.py                 | True   | 49 s
> mempool_limit.py               | True   | 16 s
> merkle_blocks.py               | True   | 25 s
> importmulti.py                 | True   | 40 s
> abandonconflict.py             | True   | 23 s
> receivedby.py                  | True   | 31 s
> mempool_resurrect_test.py      | True   | 4 s
> bip68-112-113-p2p.py           | True   | 29 s
> reindex.py                     | True   | 19 s
> txn_doublespend.py --mineblock | True   | 18 s
> rawtransactions.py             | True   | 34 s
> mempool_spendcoinbase.py       | True   | 4 s
> txn_clone.py                   | True   | 19 s
> mempool_reorg.py               | True   | 11 s
> httpbasics.py                  | True   | 11 s
> rest.py                        | True   | 21 s
> multi_rpc.py                   | True   | 4 s
> signrawtransactions.py         | True   | 5 s
> decodescript.py                | True   | 4 s
> getchaintips.py                | True   | 43 s
> proxy_test.py                  | True   | 15 s
> blockchain.py                  | True   | 7 s
> disablewallet.py               | True   | 3 s
> p2p-mempool.py                 | True   | 4 s
> keypool.py                     | True   | 11 s
> invalidblockrequest.py         | True   | 6 s
> prioritise_transaction.py      | True   | 11 s
> nodehandling.py                | True   | 26 s
> invalidtxrequest.py            | True   | 6 s
> signmessages.py                | True   | 5 s
> importprunedfunds.py           | True   | 13 s
> nulldummy.py                   | True   | 6 s
> preciousblock.py               | True   | 16 s
> p2p-versionbits-warning.py     | True   | 19 s
> rpcnamedargs.py                | True   | 4 s
> p2p-leaktests.py               | True   | 10 s
> zmq_test.py                    | True   | 19 s
> bumpfee.py                     | True   | 41 s
> listsinceblock.py              | True   | 49 s
> import-rescan.py               | True   | 59 s
>
> ALL                            | True   | 3141 s (accumulated)
>
> Runtime: 803 s
>
> Running through the extended tests and passing some options now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9956#issuecomment-287797403>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv9KnbHv9dgUbGTXuIvpv86sSmVwPks5rnpyzgaJpZM4MXbTb>
> .
>

theuni commented:
utACK 63d66ba since it works, any cleanups can happen as a next step.
